src/services_models/reserves.jl

```diff
@@ -144,6 +146,109 @@ function service_requirement_constraint!(
     return
 end
 
+_get_ramp_limits(::PSY.Component) = nothing
+_get_ramp_limits(d::PSY.ThermalGen) = PSY.get_ramp_limits(d)
+_get_ramp_limits(d::PSY.HydroGen) = PSY.get_ramp_limits(d)
+
+function _get_data_for_ramp_limit(
+    optimization_container::OptimizationContainer,
+    service::SR,
+    contributing_devices::U,
+) where {
+    SR <: PSY.Reserve,
+    U <: Union{Vector{D}, IS.FlattenIteratorWrapper{D}},
+} where {D <: PSY.Component}
+    time_frame = PSY.get_time_frame(service)
+    resolution = model_resolution(optimization_container)
+    if resolution > Dates.Minute(1)
+        minutes_per_period = Dates.value(Dates.Minute(resolution))
+    else
+        @warn("Not all formulations support under 1-minute resolutions. Exercise caution.")
+        minutes_per_period = Dates.value(Dates.Second(resolution)) / 60
+    end
+    lenght_contributing_devices = length(contributing_devices)
+    idx = 0
+    data = Vector{ServiceRampConstraintInfo}(undef, lenght_contributing_devices)
+
+    for d in contributing_devices
+        name = PSY.get_name(d)
+        non_binding_up = false
+        non_binding_down = false
+        ramp_limits = _get_ramp_limits(d)
+        if !(ramp_limits === nothing)
+            p_lims = PSY.get_active_power_limits(d)
+            max_rate = abs(p_lims.min - p_lims.max) / time_frame
+            if (ramp_limits.up >= max_rate) & (ramp_limits.down >= max_rate)
+                @debug "Generator $(name) has a nonbinding ramp limits. Constraints Skipped"
+                continue
+            else
+                idx += 1
+            end
+            ramp = (up = ramp_limits.up * time_frame, down = ramp_limits.down * time_frame)
+            data[idx] = ServiceRampConstraintInfo(name, ramp)
+        end
+    end
+    if idx < lenght_contributing_devices
+        deleteat!(data, (idx + 1):lenght_contributing_devices)
+    end
+    return data
+end
+
+function ramp_constraints!(
+    optimization_container::OptimizationContainer,
+    service::SR,
+    contributing_devices::U,
+    ::ServiceModel{SR, T},
+) where {
+    SR <: PSY.Reserve{PSY.ReserveUp},
+    T <: AbstractReservesFormulation,
+    U <: Union{Vector{D}, IS.FlattenIteratorWrapper{D}},
+} where {D <: PSY.Component}
+    initial_time = model_initial_time(optimization_container)
+    data = _get_data_for_ramp_limit(optimization_container, service, contributing_devices)
+    service_name = PSY.get_name(service)
+    if !isempty(data)
+        service_upward_rateofchange!(
+            optimization_container,
+            data,
+            make_constraint_name(RAMP, SR),
+            make_variable_name(service_name, SR),
+            service_name,
```

Comment:
the `make_constraint_name` doesn't make unique names current it will just combine the String and Service type. If you more than 1 service of type `RampReserve` you need to make multiple constraints like this, so I pass the service name and concatenate it to the constraint name. Do you have any suggestion here ?

Reply:
This is ok for now. Please add a comment about this so we can handle it when refactoring containers.